### PR TITLE
Async load reducers and persist state in parts

### PR DESCRIPTION
**Pull request by jsnajdr**
## Already merged in separate PRs

**Adding a new reducer**
Patching `combineReducers` and adding a store enhancer helps us to add a new reducer to any point in the tree:
```js
store.addReducer( [ 'extensions', 'woocommerce' ], wooReducer );
```

**Serializing any subtree into separate IndexedDB rows**
Decorate your reducer with `withStorageKey`:
```js
export default withStorageKey( 'woocommerce', wooReducer );
```
and pass it to `combineReducers` as any other reducer.

When serializing, the state will be stored into multiple IndexedDB rows:
```
redux-state-123-> root state without woo
redux-state-123:woocommerce -> the woo subtree alone
```

## Proposed in this PR

**Load and deserialize state at the right moment**
Load the state for a given `storageKey` from IndexedDB just in time.

If the `storageKey` is a part of the initial reducer, do it during store initialization. Implemented using the new `reducer.getStorageKeys` method and with multiple IndexedDB loads in `getInitialStoredState`.

The `getStateFromLocalStorage` is now used to load state pieces of various shapes. That's why it newly gets the specific reducer as a new explicit parameter. The reducer is used to deserialize the stored state.

If a reducer is added dynamically, the stored state is loaded right after the reducer is added. That's what the new `addReducerToStore` function does.

**Add reducers in extension (section) entrypoint**
Migrate the extensions' sections to this new system. The section initialization function get a new parameter (`addReducer`) that can be used to add a new reducer when the section is being loaded.

The extension reducer are removed from the initial reducer and the webpack loader that extracted reducers from sections is removed, too.

## Testing instructions
Let's test on the Zoninator extension, for example:
1. Go to `calypso.localhost:3000/?flags=-calypsoify/plugins`, disabling the redirect of Plugins UI to wp-admin.
2. Install the Zoninator extension to your Jetpack/Atomic site.
3. Go to `/extensions/zoninator/your.blog` and add a zone.
4. Reload the page.

screenshot 2018-12-03 at 19 20 01

You should see that the zone list is displayed immediately, populated from persisted state, rather than showing an intermediate placeholder while a REST request is in flight.

**Verifying the state's inner workings**
Verify that the state got persisted into two separate keys:

screenshot 2018-12-03 at 19 21 28

In console, looking at `state.extensions`, verify that the state for `state.extensions.zoninator` doesn't exist at all before you go to an `/extensions/zoninator` route.

### CODE REVIEW

**Comment by jsnajdr** [comment]
:tada: All unit tests are green for the first time today :tada: The work on the code is almost finished, now we need to review and merge all the small parts.

File: client/state/utils.js [hunk 1]
```diff
@@ -474,6 +474,40 @@ function serializeState( reducers, state, action ) {
...
+		hasChanged = hasChanged || nextStateForKey !== prevStateForKey;
+		return nextStateForKey;
+	} );
+
+	// return identical state if the stored state didn't get applied in this reducer
+	return hasChanged ? nextState : state;
+}
+
+function getStorageKeys( reducers ) {
+	return function*() {
```
Comment (flootr) [inline client/state/utils.js]: Is there a reason why we use a generator function and not e.g. an async function?

Comment (jsnajdr) [reply client/state/utils.js]: `getStorageKeys` traverses the reducer tree and gathers the `storageKey` attributes into a linear array. It's synchronous, nothing async at all.

Generator seemed to me like a very nice way to traverse the tree and concatenate values and lists of values into one big list.

Comment (jsnajdr) [reply client/state/utils.js]: I can imagine three ways of doing the tree traversal:

Generator:
```js
function*() {
  for ( const reducer of Object.values( reducers ) ) {
    if ( reducer.storageKey ) yield reducer.storageKey;
    if ( reducer.getStorageKeys ) yield* reducer.getStorageKeys();
  }
}
```

Functional version that uses `flatMap`:
```js
function() {
  return flatMap( reducers, reducer => flatMap( [
    reducer.storageKey ? [ reducer.storageKey ] : [],
    reducer.getStorageKeys ? reducer.getStorageKeys() : []
  ] ) );
}
```

Build an output array using `push` and `concat`:
```js
function() {
  let rv = [];
  for ( const reducer of Object.values( reducers ) ) {
    if ( reducer.storageKey ) rv.push( reducer.storageKey );
    if ( reducer.getStorageKeys ) rv = rv.concat( reducer.getStorageKeys() );
  }
  return rv;
}
```

I found the generator version most likeable: it's easy to read (given the reader is fluent in generators of course) and it also shouldn't allocate any ephemeral arrays (not sure if it matters performance-wise though -- maybe the Babel-transpiled generator has its own overhead that's not immediately visible).

**Review by blowery — APPROVED** [review]
LGTM. Seeing the new subkeys for zoninator. State seems to load happily.

**Pull request merged**